python/dask_cudf/dask_cudf/io/parquet.py

```diff
@@ -1,58 +1,245 @@
 # Copyright (c) 2024, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
 import functools
+import itertools
+import math
+import warnings
+from typing import TYPE_CHECKING, Any
 
+import numpy as np
 import pandas as pd
-from dask_expr.io.io import FusedParquetIO
-from dask_expr.io.parquet import FragmentWrapper, ReadParquetPyarrowFS
+from dask_expr._expr import Elemwise
+from dask_expr._util import _convert_to_list
+from dask_expr.io.io import FusedIO, FusedParquetIO
+from dask_expr.io.parquet import (
+    FragmentWrapper,
+    ReadParquetFSSpec,
+    ReadParquetPyarrowFS,
+)
 
 from dask._task_spec import Task
+from dask.dataframe.io.parquet.arrow import _filters_to_expression
+from dask.dataframe.io.parquet.core import ParquetFunctionWrapper
+from dask.tokenize import tokenize
+from dask.utils import parse_bytes
 
 import cudf
 
-from dask_cudf import _deprecated_api
+from dask_cudf import QUERY_PLANNING_ON, _deprecated_api
 
 # Dask-expr imports CudfEngine from this module
 from dask_cudf._legacy.io.parquet import CudfEngine  # noqa: F401
 
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+
+_DEVICE_SIZE_CACHE: int | None = None
+_STATS_CACHE: MutableMapping[str, Any] = {}
+
+
+def _get_device_size():
+    try:
+        # Plan A: Use PyNVML to check device size
+        import pynvml
+
+        pynvml.nvmlInit()
+        handle = pynvml.nvmlDeviceGetHandleByIndex(0)
+        return pynvml.nvmlDeviceGetMemoryInfo(handle).total
+    except ImportError:
+        # Fall back to a conservative 8GiB default
+        return "8GiB"
+
+
+def _normalize_blocksize(fraction: float = 0.03125):
```

Comment:
@ayushdg @VibhuJawa - Is this what you had in mind or is it a bit overboard?

We try finding the minimum device size in the cluster (if one is active). Otherwise we just use device 0 on the client.

Reply:
yeah, that looks good to me and lines us with what i was thinking.

Reply:
Okay, sounds good. I'm slightly worried about the scalability of querying the device size on every worker. However, that logic will be skipped if you pass in something like `blocksize="2GiB"` explicitly.

**Another Note**: This PR effectively makes `aggregate_files=True` "obsolete", because smaller files will be aggregated together at optimization time (as long as `blocksize` is not `None`). This can provide a nice drop in run time when reading many files or reading from S3 (using `aggregate_files=True` added >60s in one of the ec2 benchmarks I ran lately).